draft-ietf-httpbis-priority.md

```diff
@@ -369,70 +388,77 @@ Stream ID:
+  Length (i),
+  Prioritized Element ID (i),
+  Priority Field Value (..),
+}
 ~~~
-{: #fig-h3-reprioritization-frame title="HTTP/3 PRIORITY_UPDATE Frame Payload"}
+{: #fig-h3-reprioritization-frame title="HTTP/3 PRIORITY_UPDATE Frame"}
 
 The PRIORITY_UPDATE frame payload has the following fields:
 
-T (Prioritized Element Type):
-: A one-bit field indicating the type of element
-being prioritized. A value of 0 indicates a reprioritization for a Request
-Stream, so the Prioritized Element ID is interpreted as a Stream ID. A
-value of 1 indicates a reprioritization for a Push stream, so the Prioritized
-Element ID is interpreted as a Push ID.
-
-Empty:
-: A seven-bit field that has no semantic value.
-
 Prioritized Element ID:
 : The stream ID or push ID that is the target of the priority update.
 
 Priority Field Value:
 : The priority update value in ASCII text, encoded using Structured Headers.
 
-The HTTP/3 PRIORITY_UPDATE frame MUST NOT be sent with an invalid identifier,
-including before the request stream has been opened or before a promised
-request has been received.  If a server receives a PRIORITY_UPDATE specifying
-a push ID that has not been promised, it SHOULD be treated as a connection
-error of type H3_ID_ERROR.
-
-Because the HTTP/3 PRIORITY_UPDATE frame is sent on the control stream and
-there are no ordering guarantees between streams, a client that reprioritizes
-a request before receiving the response data might cause the server to receive
-a PRIORITY_UPDATE for an unknown request. If the request stream ID is within
-bidirectional stream limits, the PRIORITY_UPDATE frame SHOULD be buffered
-until the stream is opened and applied immediately after the request message
-has been processed. Holding PRIORITY_UPDATES consumes extra state on the peer,
-although the size of the state is bounded by bidirectional stream limits. There
-is no bound on the number of PRIORITY_UPDATES that can be sent, so an
-endpoint SHOULD store only the most recently received frame.
+The PRIORITY_UPDATE frame MAY be sent before the request stream that it
+references has been created. The Stream ID MUST be within the client-initiated
+bidirectional stream limit. If a server receives a PRIORITY_UPDATE for a Stream
+ID that is beyond the stream limits, this MUST be treated as a connection error
+of type H3_ID_ERROR. PRIORITY_UPDATE frames received before the request or
+response has started SHOULD be buffered until the stream is opened and applied
+immediately after the request message has been processed. Holding
+PRIORITY_UPDATES consumes extra state on the peer, although the size of the
+state is bounded by bidirectional stream limits. There is no bound on the number
+of PRIORITY_UPDATEs that can be sent, so an endpoint SHOULD store only the most
+recently received frame.
+
+If a server receives a PRIORITY_UPDATE specifying a Push ID that has not been
+promised, it SHOULD be treated as a connection error of type H3_ID_ERROR.
```

Comment:
This one could be a MUST.

Reply:
hmm yeah, the current text is actually a bit nasty because it doesn't restrict PRIORITY_UPDATE to be within MAX_PUSH_ID either. Combining those factors means the client has a bit too much sway over the amount of state a server needs to hold. Let me fix both things.